src/util/include/util/DispatchPool.h

```diff
@@ -0,0 +1,80 @@
+/*
+ * Xournal++
+ *
+ * Template class for a dispatch pool
+ *
+ * @author Xournal++ Team
+ * https://github.com/xournalpp/xournalpp
+ *
+ * @license GNU GPLv2 or later
+ */
+
+#pragma once
+
+#include <algorithm>
+#include <cassert>
+#include <memory>
+#include <vector>
+
+namespace xoj::util {
+template <class T>
+class DispatchPool final {
+public:
+    using listener_type = T;
+
+    template <typename... Args>
+    void dispatch(Args&&... args) const {
+        for (auto* v: this->pool) {
+            v->on(std::forward<Args>(args)...);
+        }
+    }
+
+    void add(listener_type* v) {
+        assert(v != nullptr && "Adding nullptr listener");
+        assert(std::find(this->pool.begin(), this->pool.end(), v) == this->pool.end() && "Listener is already listed");
+        this->pool.emplace_back(v);
+    }
+
+    void remove(listener_type* v) {
+        assert(v != nullptr && "Removing nullptr listener");
+        auto it = std::find(this->pool.begin(), this->pool.end(), v);
+        if (it != this->pool.end()) {
+            this->pool.erase(it);
+        }
+    }
+
+    bool empty() const { return pool.empty(); }
+
+private:
+    std::vector<listener_type*> pool;
+};
+
+template <class T>
+class Listener {
```

Comment:
Needs a doc comment.

Is my understanding of this correct: Listener is a CRTP style class. An example implementation would be `class MyListener : public Listener<MyListener> { ... }`. Subclasses of Listener are required to implement `on()` (which can potentially be overloaded), which can then be invoked by calling the `DispatchPool<MyListener>::dispatch()` method. The fun part is that because this is templated, we can abuse types like `std::enable_if_t`.

However, it seems like a concrete subclass can only implement `Listener<...>` once. Is this a problem, or do you think the overlays/views will each only be implementing one listener?

Reply:
> Is my understanding of this correct: Listener is a CRTP style class. An example implementation would be `class MyListener : public Listener<MyListener> { ... }`. Subclasses of Listener are required to implement `on()` (which can potentially be overloaded), which can then be invoked by calling the `DispatchPool<MyListener>::dispatch()` method. The fun part is that because this is templated, we can abuse types like `std::enable_if_t`.

Yes, you are correct.

> However, it seems like a concrete subclass can only implement `Listener<...>` once. Is this a problem, or do you think the overlays/views will each only be implementing one listener?

If a class needs to be two kinds of listeners, you can do that
```
class A : public Listener<A> {...};  // define virtual functions void on(...); in here
class B : pulbic Listener<B> {...};  // define virtual functions void on(...); in here - types must differ from the ones in A
class C : public A, public B {...};    // overload everything here
```
This ain't a big deal, because both `A` and `B` should be also used separately elsewhere (otherwise, don't split...).